dbt_subprojects/daily_spellbook/models/_metrics/transfers/chains/ton/metrics_ton_transfers_daily.sql

```diff
@@ -0,0 +1,105 @@
+{{ config(
+        schema = 'metrics_ton'
+        , alias = 'transfers_daily'
+        , materialized = 'incremental'
+        , file_format = 'delta'
+        , incremental_strategy = 'merge'
+        , unique_key = ['blockchain', 'block_date']
+        , incremental_predicates = [incremental_predicate('DBT_INTERNAL_DEST.block_date')]
+        )
+}}
+      
+
+with ton_prices as ( -- get price of TON for each day to estimate USD value
+    select
+        date_trunc('day', minute) as block_date
+        , avg(price) as price
+    from {{ source('prices', 'usd') }}
+    where true
+        and symbol = 'TON'
+        group by 1
+), jetton_prices as (
+    -- jetton prices based on onchain data, see https://github.com/ton-studio/dune-queries?tab=readme-ov-file#duneton_foundationresult_jetton_price_daily
+   select jp.token_address as jetton_master, ts as block_date, avg(price_usd) as price_usd
+   from dune.ton_foundation.result_jetton_price_daily jp
+   group by 1, 2
+),
+ton_flow as (
+    select block_date, source as address, -1 * value as ton_flow
+    from
+        {{ source('ton', 'messages') }}
+    where
+        direction = 'in'
+        {% if is_incremental() %}
+        and {{ incremental_predicate('block_date') }}
+        {% endif %}
+    
+    union all
+
+    select block_date, destination as address, value as ton_flow
+    from
+        {{ source('ton', 'messages') }}
+    where
+        direction = 'in'
```

Comment:
is it intended that both directions in the union here are `in`? is one supposed to be `out` to reflect the negative/positive values assigned?

Reply:
The difference between ``in`` and ``out`` messages is in the transaction related to the message. We can consider messages and transactions as a graph: messages are edges and transactions are vertices. If account A sends a messages to account B (A -> (message) -> B) we will see two entries in the ton.messages table: one with ``direction=in`` and ``tx_hash`` related to the account B and another with ``direction=out`` and ``tx_hash`` related to the account A. So in this particular case we need ``direction=in`` filter to avoid double counting of the same message.